packages/flutter/lib/src/widgets/image.dart

```diff
@@ -108,6 +109,91 @@ Future<void> precacheImage(
   return completer.future;
 }
 
+/// Signature used by [Image.frameBuilder] to control the widget that will be
+/// used when an [Image] is built.
+///
+/// The `child` argument contains the default image widget. Typically, this
+/// builder will wrap the `child` widget in some way and return the wrapped
+/// widget. If this builder returns `child` directly, it will yield the same
+/// result as if [Image.frameBuilder] was null. This argument is guaranteed to
+/// be non-null.
+///
+/// The `frame` argument specifies the index of the current image frame being
+/// rendered. It will be null before the first image frame is ready, and zero
+/// for the first image frame. For single-frame images, it will never be greater
+/// than zero. For multi-frame images (such as animated GIFs), it will increase
+/// by one every time a new image frame is shown (including when the image
+/// animates in a loop).
+///
+/// The `wasSynchronouslyLoaded` argument specifies whether the image was
+/// available synchronously (on the same
+/// [rendering pipeline frame](rendering/RendererBinding/drawFrame.html) as the
+/// `Image` widget itself was created) and thus able to be painted immediately.
+/// If this is false, then there was one or more rendering pipeline frames where
+/// the image wasn't yet available to be painted. For multi-frame images (such
+/// as animated GIFs), the value of this argument will be the same for all image
+/// frames. In other words, if the first image frame was available immediately,
+/// then this argument will be true for all image frames.
+///
+/// This builder must not return null.
+///
+/// See also:
+///
+///  * [Image.frameBuilder], which makes use of this signature in the [Image]
+///    widget.
+typedef ImageFrameBuilder = Widget Function(
+  BuildContext context,
+  Widget child,
+  int frame,
+  bool wasSynchronouslyLoaded,
+);
+
+/// Signature used by [Image.loadingBuilder] to build a representation of the
+/// image's loading progress.
+///
+/// This is useful for images that are incrementally loaded (e.g. over a local
+/// file system or a network), and the application wishes to give the user an
+/// indication of when the image will be displayed.
+///
+/// The `child` argument contains the default image widget. Typically, this
+/// builder will wrap the `child` widget in some way and return the wrapped
+/// widget. If this builder returns `child` directly, it will yield the same
+/// result as if [Image.loadingBuilder] was null. This argument is guaranteed to
+/// be non-null.
+///
+/// The `loadingProgress` argument contains the current progress towards loading
+/// the image. This argument will be non-null while the image is loading, but it
+/// will be null in the following cases:
+///
+///   * When the widget is first rendered before any bytes have been loaded.
+///   * When an image frame has been fully loaded and is available to be painted,
```

Comment:
As talked about in person, let's point out that that's not going to be a problem in the real world.

Reply:
Done